File: java/minigrep.java
/**
 * Minigrep in Java
 *
 * @author Brad White
 * @date 2025-10-21
 */
import java.io.FileInputStream;
import java.io.IOException;

class Minigrep {
  public static void main(String[] args) {
    if (args.length < 2) {
      System.out.println("Usage: minigrep <pattern> <file>");
      System.exit(1);
    }

    String pattern = args[0];
    String file = args[1];

    try {
      FileInputStream fis = new FileInputStream(file);
      byte[] buffer = new byte[1024];
      int read;
      while ((read = fis.read(buffer)) != -1) {
        String content = new String(buffer, 0, read);
        String[] lines = content.split("\n");
        boolean found = false;
        int lineNumber = 0;
        for (String line : lines) {
          if (line.contains(pattern)) {
            System.out.println(lineNumber + ": " + line);
            found = true;
          }
          lineNumber++;
        }
        if (!found) {
          System.out.println("No matches found");
        }
      }
    } catch (IOException e) {
      System.out.println("Error reading file: " + e.getMessage());
    }
  }
}

File: rust/minigrep.rs
/*
* Minigrep in Rust (functional style)
* */

fn main() {
    // Get args
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 3 {
        println!("Usage: minigrep <query> <filename>");
        std::process::exit(1);
    }
    let query = &args[1];
    let filename = &args[2];

    // Get file contents
    let contents =
        std::fs::read_to_string(filename).expect("Something went wrong reading the file");
    let lines = contents.lines();

    // Get matches
    let matches: Vec<(usize, &str)> = contents
        .lines()
        .into_iter()
        .enumerate()
        .filter(|(i, line)| line.contains(query))
        .collect();

    // print matches
    match matches.len() {
        0 => println!("No matches found for '{}' in {}", query, filename),
        _ => matches
            .into_iter()
            .for_each(|(i, line)| println!("{}: {}", i + 1, line)),
    }
}
